sct-commons/src/main/java/org/lfenergy/compas/sct/commons/LnodeTypeService.java

```diff
@@ -25,4 +28,32 @@ public Optional<TLNodeType> findLnodeType(TDataTypeTemplates tDataTypeTemplates,
         return getFilteredLnodeTypes(tDataTypeTemplates, tlNodeTypePredicate).findFirst();
     }
 
+    public Stream<DataAttributeRef> getAllDOAndDA(TDataTypeTemplates dtt, TLNodeType tlNodeType, DataAttributeRef dataRef)  {
+        return tlNodeType.getDO().stream()
+                .flatMap(tdo -> {
+                    dataRef.setLnType(tlNodeType.getId());
+                    if(tlNodeType.isSetLnClass() && !tlNodeType.getLnClass().isEmpty()) dataRef.setLnClass(tlNodeType.getLnClass().get(0));
+                    dataRef.getDoName().setName(tdo.getName());
+                    return doTypeService.findDoType(dtt, tdoType -> tdoType.getId().equals(tdo.getType()))
+                            .stream().flatMap(tdoType -> {
```

Comment:
I 'm not a big fan of using `stream` + `flatMap` on optional.  
You are twisting the Optional.  
It's better to use methods to know if you really have the value :  
```java
public Stream<DataAttributeRef> getAllDOAndDA(TDataTypeTemplates dtt, TLNodeType tlNodeType, DataAttributeRef dataRef)  {
        dataRef.setLnType(tlNodeType.getId());
        if(tlNodeType.isSetLnClass() && !tlNodeType.getLnClass().isEmpty()) dataRef.setLnClass(tlNodeType.getLnClass().get(0));
        return tlNodeType.getDO()
                .stream()
                .map(tdo -> {
                    dataRef.getDoName().setName(tdo.getName());
                    return doTypeService.findDoType(dtt, tdoType -> tdoType.getId().equals(tdo.getType()));
                })
                .filter(Optional::isPresent)
                .map(Optional::orElseThrow)
                .flatMap(tdoType -> {
                    dataRef.getDoName().setCdc(tdoType.getCdc());
                    return doTypeService.getAllSDOAndDA(dtt, tdoType, dataRef).stream();
                });
    }
```